### DISCUSSION TITLE
KubernetesPodOperator/KubernetesExecutor: Failed to adopt pod 422

### DISCUSSION THREAD

**Discussion by BillSullivan2020** [apache/airflow#23524] [General]
Hi,

I am upgrading from airflow-1.10 to airflow-2.1.4
Most of my dags use the KubernetesPodOperator to spawn worker pods for the actual logic.

As part of the upgrade, i ran into the same issue detailed here on airflow/github issue -> [21087](https://github.com/apache/airflow/issues/21087)

Based on the details in that bug-report, i tried two things.
1] to upgrade to airflow-2.3.0b1 (this will probably work if i invest more time into this / pip dependencies)
2] the source-code / workaround --> added the 5 lines of code into kubernetes_executor.py

for the #2 it seemed to fix the immediate issue, but then ran into another subsequent issue,
"Attempting to adopt pod" --> "Failed to adopt pod 422. Reason %s"

So my question is, when using KubernetesPodOperator. What if the dag failed (and some worker pod is still running).
OR, what if we restart airflow and some older worker pods are still running.
I was under the impression that KubernetesJobWatcher would try to clean-up any old-running-worker-pods that are still running in the same k8s->workspace.
However it seems that this functionality has not yet been provided/fixed/ready ?

Does airflow include this worker-pod->cleanup functionality ? or will it include this as part of the next official release ?

Thank you

**Comment by potiuk** [comment]
Airflow should adopt such tasks. There were bugs in 2.1 I think that were fixed in newer versions. I can't find the exact issues but I am quite sure task adoptiion works much better in 2.2 and 2.3

**Reply by BillSullivan2020** [reply]
Thanks for letting me know.

Next quick question.
using `airflow db init`, what is the recommended upgrade path in terms of db/airflow/version-numbers ?

e.g. airflow-1.10.x --> airflow-2.1.4 --> airflow-2.3.0 ?

right now i tried airflow-1.10.x --> airflow-2.3.0 however the db-migration has failed. (a not-null constraint was added to a column)

please advise

**Reply by potiuk** [reply]
Just follow https://airflow.apache.org/docs/apache-airflow/stable/upgrading-from-1-10/index.html . You can migrate to any version.  There are no recommendations to do it differently. If you had problem with migrations they lln you should create an issue describing your circumstances in detail. It might be a bug or nlunsupported db (check prerequisites) or maybe some manual modification of your db or simply some old data. 

You need to provide all details in an issue - but do it after you follow the migration guide meticulously step by step. There are some upgrade checks there that might reveal some issues. 

Just follow.the docs

**Comment by BillSullivan2020** [comment]
Hi. this post's original issue, "Failed to adopt pod XYZ. 422 Unprocessable entity". i have tested on airflow-2.3.0 and the bug still persists. i.e. it does not clearDown orphanWorkerPods.
Any idea when this is going to be fixed ?
Thanks for your time.

**Reply by potiuk** [reply]
I think the issue you have migh be looked at when you create an issue and describe enough logs and information so that someone might be able to investigate and fix it. By providing detailed, reproiducible case you increase the chances that someone will actually fix it.

So the answer is - the more you provide helpful information on the issue, the sooner it will be fixed. 

There is really no point in saying "issue still persists" when - apparently - the original issue has been fixed. It's a bit lazy and in most cases  you have no ground to say your issue is "the same". By saying "issue still persist" and not providing all the details you add exactly 0 value. 


If you actually "know" that your issue is "the same" (it might have similar error but might be different one), then likely you tracked it down and you know exactly what causes it and how to fix it - so better to make a PR for that and fix it (including adding a test case). 

So please - open an issue with enough details for the maintainers to be able to fix the issue - even if you **think** it's the same issue.

**Comment by BillSullivan2020** [comment]
Hiya, PotiUK.

Thanks for your reply.

Let me tell you how to reproduce this.

Goal of this: to describe how to reproduce the "Failed to Adopt pod" error condition.

The DAG->step Described Below should be of type KubernetesPodOperator

NOTE: under normal operation,
                             (where the MAIN_AIRFLOW_POD is never recycled by k8s, we will never see this edge-case)
                             (it is only when the workerPod is still running, but the MAIN_AIRFLOW_POD is suddenly restarted/stopped)
                             (that we would see orphan->workerPods)

1. Implement a contrived-DAG, with a single step -> which is long-running (e.g. 6 minutes)
2. Deploy your airflow-2.1.4 / airfow-2.3.0 together with the contrived-DAG
3. Run your contrived-DAG.
4. in the middle of running the single-step, check via "kubectl" that your Kubernetes->workerPod has been created / running
5. while workerPod still running, do "kubectl delete pod <OF_MAIN_AIRFLOW_POD>". This will mean that the workerPod becomes an orphan.
6. the workerPod still continues to run through to completion. after which the K8S->status of the pod will be Completed, however the pod doesn't shut down itself.
7. "kubectl" start up a new <MAIN_AIRFLOW_POD> so the web-ui is running again.
8. MAIN_AIRFLOW_POD->webUi - Run your contrived-DAG again
9. while the contrived-DAG is starting/tryingToStart etc, you will see in the logs printed out "Failed to adopt pod" -> with 422 error code.

The step-9 with the error message, you will find two appearances of this error msg in the airflow-2.1.4, airflow-2.3.0 source-code.
The step-7 may also - general logging from the MAIN_APP - may also output the "Failed to adopt pod" error message also.

**Reply by BillSullivan2020** [reply]
interestingly -> adding a bit of information for you. The old airflow-1.10.x series does not exhibit this issue with the orphan pods; instead, it correctly adopts/handles/cleanUp those old worker pods, even after an abrupt restart of the MAIN_POD.

only seems to be the later versions (i have seen this on both 2.1.4 & 2.3.0) whereby it cannot adopt orphan pods, with the above error message.

**Reply by potiuk** [reply]
Can you please create an issue from that @BillSullivan2020 ? Since you have a reproducible issue and detailed reproduction steps you should open an issue for it. There are a number things to fill there and it will make someone who is closer to Kubernetes to take a look at it. Then also posting gists to the logs you have would make it easier to analyze it.

**Reply by potiuk** [reply]
Just so you understand - discussion is a discussion to find out if something is a bug when you are not sure it's an airflow issue. But you seem to have a detailed reproduction steps and whenever we diagnose or fix it, those who will be looking thee will likely ask you for addtional information, more logs or veriification if workarounds might work - that's why you should open an issue for that.

**Reply by BillSullivan2020** [reply]
Hello Potiuk, thank you for your feedback! Please find here the issue/bug that i created. https://github.com/apache/airflow/issues/24015
